crt0stack/src/lib.rs

```diff
@@ -0,0 +1,270 @@
+            AuxvEntry::Base(v) => self.auxv.push(AuxvPair {
+                key: AT_BASE,
+                value: v as _,
+            }),
+            AuxvEntry::Flags(v) => self.auxv.push(AuxvPair {
+                key: AT_FLAGS,
+                value: v as _,
+            }),
+            AuxvEntry::Entry(v) => self.auxv.push(AuxvPair {
+                key: AT_ENTRY,
+                value: v as _,
+            }),
+            AuxvEntry::Notelf(v) => self.auxv.push(AuxvPair {
+                key: AT_NOTELF,
+                value: v as _,
+            }),
+            AuxvEntry::Uid(v) => self.auxv.push(AuxvPair {
+                key: AT_UID,
+                value: v as _,
+            }),
+            AuxvEntry::EUid(v) => self.auxv.push(AuxvPair {
+                key: AT_EUID,
+                value: v as _,
+            }),
+            AuxvEntry::Gid(v) => self.auxv.push(AuxvPair {
+                key: AT_GID,
+                value: v as _,
+            }),
+            AuxvEntry::EGid(v) => self.auxv.push(AuxvPair {
+                key: AT_EGID,
+                value: v as _,
+            }),
+            AuxvEntry::Platform(v) => {
+                self.platform.replace(v);
+            }
+            AuxvEntry::HWCap(v) => self.auxv.push(AuxvPair {
+                key: AT_HWCAP,
+                value: v as _,
+            }),
+            AuxvEntry::ClockTick(v) => self.auxv.push(AuxvPair {
+                key: AT_CLKTCK,
+                value: v as _,
+            }),
+            AuxvEntry::Secure(v) => self.auxv.push(AuxvPair {
+                key: AT_SECURE,
+                value: if v { 1 } else { 0 },
+            }),
+            AuxvEntry::BasePlatform(v) => {
+                self.base_platform.replace(v);
+            }
+            AuxvEntry::Random(v) => {
+                self.random.replace(v);
+            }
+            AuxvEntry::HWCap2(v) => self.auxv.push(AuxvPair {
+                key: AT_HWCAP2,
+                value: v as _,
+            }),
+            AuxvEntry::ExecFilename(v) => {
+                self.exec_fn.replace(v);
+            }
+        }
+    }
+
+    /// serialize to a mutable byte slice
+    ///
+    ///
+    pub fn serialize<'a>(self, slice: &mut [u8]) -> Result<usize, Error> {
+        let mut sp = Stack::new(slice);
+        let mut str_ptr = Vec::new();
+        let mut rand_ptr: Option<*const u8> = None;
+        let mut platform_ptr: Option<*const u8> = None;
+        let mut base_platform_ptr: Option<*const u8> = None;
+        let mut exec_fn_ptr: Option<*const u8> = None;
+
+        str_ptr.push(core::ptr::null());
+        for s in self.envp {
+            sp.push_l(0u8);
+            sp.push_slice(s.as_bytes());
+            str_ptr.push(sp.ptr());
+        }
+        str_ptr.push(core::ptr::null());
+        for s in self.argv.iter().rev() {
+            sp.push_l(0u8);
+            sp.push_slice(s.as_bytes());
+            str_ptr.push(sp.ptr());
+        }
+
```

Comment:
Do we need to pad to a 16B boundary here?

Referring to this component that follows the args on the diagram for the lwn article that Nathaniel linked:

```
           /   4fd4: 'two\0'                                     |       | |     <----+
     args |    4fd0: 'one\0'                                     |       | |    <---+ |
           \_  4fcb: 'zero\0'                                    |       | |   <--+ | |
               3020: random gap padded to 16B boundary           |       | |      | | |
    - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -|       | |      | | |
               3019: 'x86_64\0'                        <-+       |       | |      | | 
```

Reply:
added padding